milestone 1: eda

digma

### imports and configuration

In [6]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

### load finalized datasets

In [11]:
demographics = pd.read_csv('../data/processed/customer_demographics_preprocessed.csv')
transactions = pd.read_csv('../data/processed/customer_transactions_preprocessed.csv')
interactions = pd.read_csv('../data/processed/social_media_interactions_preprocessed.csv')

print("Customer Demographics:", demographics.shape)
print("Customer Transactions:", transactions.shape)
print("Social Media Interactions:", interactions.shape)

Customer Demographics: (3000, 6)
Customer Transactions: (3015, 6)
Social Media Interactions: (3020, 6)


###  convert temporal variables

In [13]:
demographics["SignupDate"] = pd.to_datetime(
    demographics["SignupDate"],
    errors="coerce"
)

transactions["TransactionDate"] = pd.to_datetime(
    transactions["TransactionDate"],
    errors="coerce"
)

interactions["InteractionDate"] = pd.to_datetime(
    interactions["InteractionDate"],
    errors="coerce"
)

###  verify dataset structure

In [14]:
print("=== CUSTOMER DEMOGRAPHICS ===")
print(demographics.info())
print("\nShape:", demographics.shape)
print("\nColumns:")
print(demographics.columns.tolist())

print("\n=== CUSTOMER TRANSACTIONS ===")
print(transactions.info())
print("\nShape:", transactions.shape)
print("\nColumns:")
print(transactions.columns.tolist())

print("\n=== SOCIAL MEDIA INTERACTIONS ===")
print(interactions.info())
print("\nShape:", interactions.shape)
print("\nColumns:")
print(interactions.columns.tolist())

=== CUSTOMER DEMOGRAPHICS ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   CustomerID   3000 non-null   object        
 1   Age          3000 non-null   float64       
 2   Gender       3000 non-null   object        
 3   Location     3000 non-null   object        
 4   IncomeLevel  3000 non-null   object        
 5   SignupDate   3000 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(1), object(4)
memory usage: 140.8+ KB
None

Shape: (3000, 6)

Columns:
['CustomerID', 'Age', 'Gender', 'Location', 'IncomeLevel', 'SignupDate']

=== CUSTOMER TRANSACTIONS ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3015 entries, 0 to 3014
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   CustomerID       3015 non-null   object      

### post preprocecssing validation

In [15]:
datasets = {
    "Customer Demographics": demographics,
    "Customer Transactions": transactions,
    "Social Media Interactions": interactions
}

for name, df in datasets.items():
    print(f"\n{'=' * 60}")
    print(name)
    print(f"{'=' * 60}")

    print("\nShape:")
    print(df.shape)

    print("\nMissing values:")
    print(df.isna().sum())

    print("\nExact duplicate rows:")
    print(df.duplicated().sum())


Customer Demographics

Shape:
(3000, 6)

Missing values:
CustomerID     0
Age            0
Gender         0
Location       0
IncomeLevel    0
SignupDate     0
dtype: int64

Exact duplicate rows:
0

Customer Transactions

Shape:
(3015, 6)

Missing values:
CustomerID         0
TransactionID      0
TransactionDate    0
Amount             0
ProductCategory    0
PaymentMethod      0
dtype: int64

Exact duplicate rows:
0

Social Media Interactions

Shape:
(3020, 6)

Missing values:
CustomerID         0
InteractionID      0
InteractionDate    0
Platform           0
InteractionType    0
Sentiment          0
dtype: int64

Exact duplicate rows:
0


In [16]:
print("CustomerID duplicates:",
      demographics["CustomerID"].duplicated().sum())

print("TransactionID duplicates:",
      transactions["TransactionID"].duplicated().sum())

print("InteractionID duplicates:",
      interactions["InteractionID"].duplicated().sum())

CustomerID duplicates: 0
TransactionID duplicates: 15
InteractionID duplicates: 20


### customer demographics eda

####  numerical descriptive statistics

In [17]:
age_stats = demographics["Age"].describe()

print("Age Descriptive Statistics")
display(age_stats)

print("\nMean:", demographics["Age"].mean())
print("Median:", demographics["Age"].median())
print("Mode:", demographics["Age"].mode().tolist())
print("Minimum:", demographics["Age"].min())
print("Maximum:", demographics["Age"].max())
print("Range:", demographics["Age"].max() - demographics["Age"].min())
print("Variance:", demographics["Age"].var())
print("Standard Deviation:", demographics["Age"].std())

Age Descriptive Statistics


count   3,000.00
mean       44.66
std        14.45
min        18.00
25%        33.75
50%        45.00
75%        57.00
max        70.00
Name: Age, dtype: float64


Mean: 44.66433333333333
Median: 45.0
Mode: [45.0]
Minimum: 18.0
Maximum: 70.0
Range: 52.0
Variance: 208.88395587418222
Standard Deviation: 14.452818267527695


####  gender distribution

In [20]:
gender_distribution = pd.DataFrame({
    "Count": demographics["Gender"].value_counts(),
    "Percentage": demographics["Gender"].value_counts(normalize=True) * 100
})

display(gender_distribution)

,Count,Percentage
Gender,,
Female,1515,50.50
Male,1485,49.50


####  income distribution

In [21]:
income_distribution = pd.DataFrame({
    "Count": demographics["IncomeLevel"].value_counts(),
    "Percentage": demographics["IncomeLevel"].value_counts(normalize=True) * 100
})

display(income_distribution)

,Count,Percentage
IncomeLevel,,
High,935,31.17
Low,918,30.60
Medium,863,28.77
Unknown,284,9.47


####  signup date coverage

In [23]:
print("Earliest signup date:", demographics["SignupDate"].min())
print("Latest signup date:", demographics["SignupDate"].max())

signup_year_distribution = (
    demographics["SignupDate"]
    .dt.year
    .value_counts()
    .sort_index()
)

display(signup_year_distribution)

Earliest signup date: 2019-07-01 00:00:00
Latest signup date: 2024-06-30 00:00:00


SignupDate
2019    296
2020    599
2021    590
2022    571
2023    613
2024    331
Name: count, dtype: int64

###  customer transactions eda

####  amount statistics

In [24]:
amount_stats = transactions["Amount"].describe()

print("Transaction Amount Descriptive Statistics:")
display(amount_stats)

print("\nMean:", transactions["Amount"].mean())
print("Median:", transactions["Amount"].median())
print("Mode:", transactions["Amount"].mode().tolist())
print("Minimum:", transactions["Amount"].min())
print("Maximum:", transactions["Amount"].max())
print("Range:", transactions["Amount"].max() - transactions["Amount"].min())
print("Variance:", transactions["Amount"].var())
print("Standard Deviation:", transactions["Amount"].std())

Transaction Amount Descriptive Statistics:


count   3,015.00
mean      499.29
std       274.89
min         0.00
25%       269.85
50%       504.14
75%       720.39
max       999.86
Name: Amount, dtype: float64


Mean: 499.2938905472637
Median: 504.14
Mode: [504.14]
Minimum: 0.0
Maximum: 999.86
Range: 999.86
Variance: 75562.47302344492
Standard Deviation: 274.8862910795024


####  product category distribution

In [25]:
product_distribution = pd.DataFrame({
    "Count": transactions["ProductCategory"].value_counts(),
    "Percentage": transactions["ProductCategory"].value_counts(normalize=True) * 100
})

display(product_distribution)

,Count,Percentage
ProductCategory,,
Clothing,571,18.94
Electronics,564,18.71
Automotive,540,17.91
Home & Garden,530,17.58
Health & Beauty,528,17.51
Unknown,282,9.35


####  payment method distribution

In [27]:
payment_distribution = pd.DataFrame({
    "Count": transactions["PaymentMethod"].value_counts(),
    "Percentage": transactions["PaymentMethod"].value_counts(normalize=True) * 100
})

display(payment_distribution)

,Count,Percentage
PaymentMethod,,
Debit Card,767,25.44
Credit Card,755,25.04
PayPal,750,24.88
Bank Transfer,743,24.64


####  product category vs amount

In [29]:
product_amount_analysis = (
    transactions
    .groupby("ProductCategory")["Amount"]
    .agg(["count", "mean", "median", "std"])
    .sort_values("mean", ascending=False)
)

display(product_amount_analysis)

,count,mean,median,std
ProductCategory,,,,
Home & Garden,530,522.74,504.14,272.46
Automotive,540,502.74,504.14,272.41
Clothing,571,501.06,504.14,280.23
Electronics,564,495.25,504.14,264.66
Unknown,282,483.86,504.14,274.45
Health & Beauty,528,482.89,504.14,284.31


Interpretation: Home & Garden has the highest observed mean, while Health & Beauty has the lowest. However, these are descriptive differences, not evidence of statistical significance.

####  payment method vs amount 

In [30]:
payment_amount_analysis = (
    transactions
    .groupby("PaymentMethod")["Amount"]
    .agg(["count", "mean", "median", "std"])
    .sort_values("mean", ascending=False)
)

display(payment_amount_analysis)

,count,mean,median,std
PaymentMethod,,,,
Credit Card,755,507.18,504.14,271.03
PayPal,750,503.18,504.14,270.45
Debit Card,767,499.82,504.14,278.53
Bank Transfer,743,486.82,504.14,279.52


####  transaction temporal pattern

In [31]:
monthly_transactions = (
    transactions
    .groupby(transactions["TransactionDate"].dt.to_period("M"))
    .agg(
        TransactionCount=("TransactionID", "count"),
        AverageAmount=("Amount", "mean"),
        TotalAmount=("Amount", "sum")
    )
)

display(monthly_transactions)

,TransactionCount,AverageAmount,TotalAmount
TransactionDate,,,
2022-07,119,513.49,"61,105.00"
2022-08,113,472.83,"53,429.68"
2022-09,120,472.18,"56,661.21"
2022-10,119,504.16,"59,994.77"
2022-11,126,479.54,"60,422.16"
2022-12,135,532.37,"71,869.59"
2023-01,119,460.27,"54,772.60"
2023-02,110,470.40,"51,743.67"
2023-03,135,517.58,"69,873.68"


there is variation in monthly transaction volume but no obvious monotonic upward or downward trend

###  social media interactions eda

####  platform

In [32]:
platform_distribution = pd.DataFrame({
    "Count": interactions["Platform"].value_counts(),
    "Percentage": interactions["Platform"].value_counts(normalize=True) * 100
})

display(platform_distribution)

,Count,Percentage
Platform,,
Instagram,921,30.50
Twitter,917,30.36
Facebook,891,29.50
Unknown,291,9.64


####  interaction type

In [33]:
interaction_type_distribution = pd.DataFrame({
    "Count": interactions["InteractionType"].value_counts(),
    "Percentage": interactions["InteractionType"].value_counts(normalize=True) * 100
})

display(interaction_type_distribution)

,Count,Percentage
InteractionType,,
Comment,1018,33.71
Share,1007,33.34
Like,995,32.95


####  sentiment

In [35]:
sentiment_distribution = pd.DataFrame({
    "Count": interactions["Sentiment"].value_counts(),
    "Percentage": interactions["Sentiment"].value_counts(normalize=True) * 100
})

display(sentiment_distribution)

,Count,Percentage
Sentiment,,
Positive,905,29.97
Neutral,874,28.94
Negative,866,28.68
Unknown,309,10.23
Very Negative,34,1.13
Very Positive,32,1.06


The main three sentiment classes are balanced, but:

10.23% of interactions have Unknown sentiment.

And the extreme sentiment classes are rare:

Very Negative = 1.13%
Very Positive = 1.06%

####  social media interactions temporal analysis

In [36]:
monthly_interactions = (
    interactions
    .groupby(interactions["InteractionDate"].dt.to_period("M"))
    .size()
    .rename("InteractionCount")
)

display(monthly_interactions)

InteractionDate
2023-07    273
2023-08    242
2023-09    242
2023-10    266
2023-11    233
2023-12    259
2024-01    259
2024-02    237
2024-03    251
2024-04    254
2024-05    264
2024-06    240
Freq: M, Name: InteractionCount, dtype: int64

Actual coverage:

July 1, 2023 – June 30, 2024

Monthly interaction volume ranges from:

233 to 273 interactions

Again, there is variation but no obvious monotonic trend.

## cross-dataset analysis

In [39]:
# since the three datasets have different grains, we must not directly correlate individual transaction rows with individual customer demographic rows
## first, aggregate transactions and interactions to the customer level

transaction_customer_summary = (
    transactions
    .groupby("CustomerID")
    .agg(
        TotalSpend=("Amount", "sum"),
        TransactionCount=("TransactionID", "count"),
        AverageTransactionAmount=("Amount", "mean")
    )
    .reset_index()
)

interactions_customer_summary = (
    interactions
    .groupby("CustomerID")
    .agg(
        InteractionCount=("InteractionID", "count")
    )
    .reset_index()
)

customer_analysis = (
    demographics
    .merge(transaction_customer_summary, on="CustomerID", how="left")
    .merge(interactions_customer_summary, on="CustomerID", how="left")
)

customer_analysis[
    ["TotalSpend", "TransactionCount",
     "AverageTransactionAmount", "InteractionCount"]
] = customer_analysis[
    ["TotalSpend", "TransactionCount",
     "AverageTransactionAmount", "InteractionCount"]
].fillna(0)

display(customer_analysis.head())

,CustomerID,Age,Gender,Location,IncomeLevel,SignupDate,TotalSpend,TransactionCount,AverageTransactionAmount,InteractionCount
0,0009fdd2-ae63-45ca-8d5b-d0ea98381f7b,21.00,Female,Lake George,Low,2020-11-09,389.69,1.00,389.69,2.00
1,000c6bbd-533a-432d-922c-ab64197e71c5,25.00,Male,North Oliviaton,High,2019-11-06,0.00,0.00,0.00,4.00
2,00115fc0-f155-42cd-ba68-58aab67b3360,36.00,Male,Nelsonstad,Low,2022-02-16,445.48,2.00,222.74,0.00
3,00145374-004a-4685-af4d-c8a8967b969e,69.00,Female,Jacquelinehaven,Unknown,2023-04-28,0.00,0.00,0.00,2.00
4,0047d8ce-a5bb-4db7-860f-6ff66e1cd060,40.00,Male,Mitchellview,Medium,2021-11-11,259.47,1.00,259.47,0.00


### cross-dataset correlations

In [40]:
customer_numeric = customer_analysis[
    [
        "Age",
        "TotalSpend",
        "TransactionCount",
        "AverageTransactionAmount",
        "InteractionCount"
    ]
]

correlation_matrix = customer_numeric.corr(method="pearson")

display(correlation_matrix)

,Age,TotalSpend,TransactionCount,AverageTransactionAmount,InteractionCount
Age,1.00,-0.00,-0.02,0.01,0.01
TotalSpend,-0.00,1.00,0.88,0.80,-0.01
TransactionCount,-0.02,0.88,1.00,0.60,-0.01
AverageTransactionAmount,0.01,0.80,0.60,1.00,0.00
InteractionCount,0.01,-0.01,-0.01,0.00,1.00


The Age–TotalSpend relationship is essentially zero in this dataset.

The InteractionCount–TotalSpend relationship is also essentially zero.

The strong TransactionCount–TotalSpend relationship should not be presented as a discovery that transactions cause spending, because TotalSpend is mathematically constructed by summing transaction amounts and TransactionCount is derived from the same transaction records. That relationship is structurally expected.

### anova: product category vs amount

In [41]:
from scipy.stats import f_oneway

product_groups = [
    group["Amount"].values
    for _, group in transactions.groupby("ProductCategory")
]

anova_result = f_oneway(*product_groups)

print("F-statistic:", anova_result.statistic)
print("p-value:", anova_result.pvalue)

F-statistic: 1.3724165735129192
p-value: 0.23144138659618788


At α = 0.05:

Since p = 0.2314 > 0.05, fail to reject H₀. The available data does not provide sufficient statistical evidence of a difference in mean transaction Amount across ProductCategory groups.